File: 프로그래머스/3/161988. 연속 펄스 부분 수열의 합/연속 펄스 부분 수열의 합.py
def solution(sequence):
    max_sum = 0

    curr1 = 0  # [1, -1, 1, -1 ...]
    curr2 = 0  # [-1, 1, -1, 1 ...]

    for i, num in enumerate(sequence):
        sign1 = 1 if i % 2 == 0 else -1
        sign2 = -sign1

        curr1 = max(num * sign1, curr1 + num * sign1)
        curr2 = max(num * sign2, curr2 + num * sign2)

        max_sum = max(max_sum, curr1, curr2)

    return max_sum

File: 프로그래머스/3/161988. 연속 펄스 부분 수열의 합/연속 펄스 부분 수열의 합.kt
class Solution {
    fun solution(sequence: IntArray): Long {
        var maxSum = 0L

        var curr1 = 0L // [1, -1, 1, -1 ...]
        var curr2 = 0L // [-1, 1, -1, 1 ...]

        for (i in sequence.indices) {
            val sign1 = if (i % 2 == 0) 1 else -1
            val sign2 = -sign1

            val v1 = sequence[i].toLong() * sign1
            val v2 = sequence[i].toLong() * sign2

            curr1 = maxOf(v1, curr1 + v1)
            curr2 = maxOf(v2, curr2 + v2)

            maxSum = maxOf(maxSum, curr1, curr2)
        }

        return maxSum
    }
}
